x/mongo/driver/operation.go

```diff
@@ -1078,6 +1094,27 @@ func (op Operation) addServerAPI(dst []byte) []byte {
 	return dst
 }
 
+// addMaxTimeMS adds a 'maxTimeMS' field to the wire message in dst based on the current context's deadline
+// and the 90th percentile RTT if op.Timeout is set.
+func (op Operation) addMaxTimeMS(ctx context.Context, dst []byte, desc description.SelectedServer) []byte {
+	if op.Timeout != nil {
+		// Retrieve current deadline from context, convert it to timeout and append
+		// maxTimeMS as (timeout - 90th percentile RTT).
+		if deadline, deadlineSet := ctx.Deadline(); deadlineSet {
+			remainingTimeout := time.Until(deadline)
+			// If 90th percentile RTT is less than remainingTimeout, error instead
+			// of attempting to send message to server.
+			// TODO(CSOT): Uncomment this once RTT90 is actually being calculated.
+			// if desc.RTT90 < remainingTimeout {
+			//	return ErrDeadlineWouldBeExceeded
```

Comment:
At this point of error return, `op.Timeout != nil` because of L1100, so the source is always `Timeout`. The appension of `maxTimeMS` does not happen if `op.Timeout == nil` and there is a deadline on the provided context (users must opt-in to the behavior by specifying a `Timeout` value on `Client`, `Database` or `Collection`). I did add some more context to the error as described above with `internal.WrapErrorf`.